### PR TITLE
Image ocr fixes

### PR DESCRIPTION
**Pull request by mbhenry**
This PR fixes the problems people have been having with image_ocr.py converging on more recent versions of Keras and Tensorflow.  I tested this on the master branch of Keras with Tensorflow 0.11.0 and Theano 0.9.0 and consistently achieve the normalized edit distances mentioned in the docstring. 

The key to fixing it was starting with images that are 128 pixels wide instead of 512.  The width corresponds to the "temporal" domain in the RNN, so narrower images means less backprop through time and convergence seems to be much easier.  After 20 epochs, the model is rebuilt for 512 wide images with the same weights and the network has no problem converging on longer inputs and more words.  This PR also makes the curriculum learning more fine grained in the first 12 epochs by rebinding the text paint function to make it increasingly more challenging. This leads to much faster convergence than before. 

Here is the output on Epoch 10:

e10

and Epoch 23:

e23

### CODE REVIEW

File: examples/image_ocr.py [hunk 1]
```diff
@@ -79,28 +74,33 @@ def speckle(img):
 # also uses a random font, a slight random rotation,
 # and a random amount of speckle noise
 
-def paint_text(text, w, h):
+def paint_text(text, w, h, rotate=False, ud=False, multi_fonts=False):
     surface = cairo.ImageSurface(cairo.FORMAT_RGB24, w, h)
```
Comment (fchollet) [inline examples/image_ocr.py]: Not sure what the cairo API does, but with Keras / TF / Theano conventions, `height` comes before `width`.

File: examples/image_ocr.py [hunk 2]
```diff
@@ -229,36 +245,35 @@ def build_word_list(self, num_words, max_string_len=None, mono_fraction=0.5):
     # painting of the text is performed
     def get_batch(self, index, size, train):
         if K.image_dim_ordering() == 'th':
-            X_data = np.ones([size, 1, self.img_h, self.img_w])
+            X_data = np.ones([size, 1, self.img_w, self.img_h])
```
Comment (fchollet) [inline examples/image_ocr.py]: With Keras / TF / Theano conventions, `height` comes before `width`.

File: examples/image_ocr.py [hunk 3]
```diff
@@ -362,109 +377,118 @@ def show_edit_distance(self, num):
...
+                                 img_w=img_w,
+                                 img_h=img_h,
+                                 downsample_factor=(pool_size ** 2),
+                                 val_split=words_per_epoch - val_words,
+                                 start_epoch=start_epoch
+                                 )
+    act = 'relu'
+    input_data = Input(name='the_input', shape=input_shape, dtype='float32')
+    inner = Convolution2D(conv_num_filters, filter_size, filter_size, border_mode='same',
+                          activation=act, init="he_normal", name='conv1')(input_data)
```
Comment (fchollet) [inline examples/image_ocr.py]: Are you sure about these choices of initialization? The default should work well. Also you are using inconsistent quote characters (should be `'` everywhere).

File: examples/image_ocr.py [hunk 4]
```diff
@@ -362,109 +377,118 @@ def show_edit_distance(self, num):
...
+    # Two layers of bidirecitonal GRUs
+    # GRU seems to work as well, if not better than LSTM:
+    gru_1 = GRU(rnn_size, return_sequences=True, init="he_normal", name='gru1')(inner)
+    gru_1b = GRU(rnn_size, return_sequences=True, go_backwards=True, init="he_normal", name='gru1_b')(inner)
+    gru1_merged = merge([gru_1, gru_1b], mode='sum')
+    gru_2 = GRU(rnn_size, return_sequences=True, init="he_normal", name='gru2')(gru1_merged)
+    gru_2b = GRU(rnn_size, return_sequences=True, go_backwards=True, init="he_normal", name='gru2_b')(gru1_merged)
+
+    # transforms RNN output to character activations:
+    inner = TimeDistributed(Dense(img_gen.get_output_size(), init="he_normal",
```
Comment (fchollet) [inline examples/image_ocr.py]: You can simply use a Dense layer here.

**Comment by mbhenry** [comment]
The reason I switched width/height was it saves me a permute layer since the width dimension has to be the "time" dimension into the RNN. Would you want to have the permute back in place in the interest of following convention?

The he_normal does seem to speed up initial convergence:

| Epoch | he_normal logloss | default (glorot_uniform) logloss |
| --- | --- | --- |
| 1 | 9.3425 | 11.0996 |
| 2 | 0.5780 | 5.9318 |
| 3 | 0.0184 | 0.3271 |

I can submit a fix for the quotes/Dense thing once I get guidance on how to handle the width/height/permute question.

**Comment by fchollet** [comment]
About width/height: you can leave it as is, but please add a comment in the code to clarify. It might confuse some people.

**Comment by patyork** [comment]
This is a significant improvement over the older one - very nice job.

File: examples/image_ocr.py [hunk 5]
```diff
@@ -362,109 +377,118 @@ def show_edit_distance(self, num):
...
+
+    viz_cb = VizCallback(run_name, start_epoch, test_func, img_gen.next_val())
+
+    model.fit_generator(generator=img_gen.next_train(), samples_per_epoch=(words_per_epoch - val_words),
+                        nb_epoch=num_epochs, validation_data=img_gen.next_val(), nb_val_samples=val_words,
+                        callbacks=[viz_cb, img_gen])
+
+
+if __name__ == "__main__":
+    run_name = datetime.datetime.now().strftime('%Y:%m:%d:%H:%M:%S')
```
Comment (patyork) [inline examples/image_ocr.py]: I would change this back to `datetime.datetime.now().strftime('%A, %d. %B %Y %I.%M%p'))` for better formatting.

File: examples/image_ocr.py [hunk 6]
```diff
@@ -362,109 +379,118 @@ def show_edit_distance(self, num):
...
+
+    # clipnorm seems to speeds up convergence
+    sgd = SGD(lr=0.02, decay=1e-6, momentum=0.9, nesterov=True, clipnorm=5)
+
+    model = Model(input=[input_data, labels, input_length, label_length], output=[loss_out])
+
+    # the loss calc occurs elsewhere, so use a dummy lambda func for the loss
+    model.compile(loss={'ctc': lambda y_true, y_pred: y_pred}, optimizer=sgd)
+    if start_epoch > 0:
+        weight_file = os.path.join(OUTPUT_DIR, os.path.join(run_name, 'weights%0d.h5' % (start_epoch - 1)))
```
Comment (patyork) [inline examples/image_ocr.py]: The loaded file formatting needs to be `'weights%02d.h5'` to match the save portion on #L382 else will fail for start epochs <=10

Comment (mbhenry) [inline examples/image_ocr.py]: Good catch!   That reloading model thing was a hacky afterthought since I could not get variable width inputs to work.

File: examples/image_ocr.py [hunk 7]
```diff
@@ -362,109 +379,118 @@ def show_edit_distance(self, num):
...
+    model.compile(loss={'ctc': lambda y_true, y_pred: y_pred}, optimizer=sgd)
+    if start_epoch > 0:
+        weight_file = os.path.join(OUTPUT_DIR, os.path.join(run_name, 'weights%02d.h5' % (start_epoch - 1)))
+        model.load_weights(weight_file)
+    # captures output of softmax so we can decode the output during visualization
+    test_func = K.function([input_data], [y_pred])
+
+    viz_cb = VizCallback(run_name, start_epoch, test_func, img_gen.next_val())
+
+    model.fit_generator(generator=img_gen.next_train(), samples_per_epoch=(words_per_epoch - val_words),
```
Comment (patyork) [inline examples/image_ocr.py]: The fit_generator now takes an intiial_epoch parameter (a surprisingly recent addition?). I think the `fit_gen` should reflect that with `initial_epoch=start_epoch`. Not sure that it will affect much, but some possible issues are:

- No matter the start epoch, it will run the given `num_epochs`
- The data generator may revert to the first stages of learning (simplified inputs) since it thinks its going through epoch 0, 1, 2, 3, etc..?

**Comment by patyork** [comment]
Sorry for the multiple random comments. Just running the script and letting any issue pop up as I run into restarts and what not.

File: examples/image_ocr.py (hunk 7 above)
Comment (mbhenry) [inline examples/image_ocr.py]: Didn't notice that initial_epoch thing...seems to be buried in the source, but the intent according to comments is exactly what this example was doing. 

The data generator is passed the start_epoch to avoid going back to those simplified inputs, and the num_epochs is passed to the train function. I think using initial_epoch will be cleaner, so I can try it tomorrow AM.

Comment (patyork) [inline examples/image_ocr.py]: Sure. I just noticed you are handling it on line 307 yourself, but it doesn't flow through everywhere.

File: examples/image_ocr.py [hunk 8]
```diff
@@ -286,20 +303,18 @@ def next_val(self):
...
-            self.build_word_list(64000, 12, 0.5)
+        cur_epoch = epoch + self.start_epoch
+        # rebind the paint function to implement curriculum learning
+        if cur_epoch == 3:
+            self.paint_func = lambda text: paint_text(text, self.img_w, self.img_h,
+                                                      rotate=False, ud=True, multi_fonts=False)
+        if cur_epoch == 6:
+            self.paint_func = lambda text: paint_text(text, self.img_w, self.img_h,
+                                                      rotate=False, ud=True, multi_fonts=True)
+        if cur_epoch == 12:
```
Comment (patyork) [inline examples/image_ocr.py]: When restarting (from scratch) using saved weights, no matter `start_epoch`, the training data starts with the simplest data, until it hits one of these comparisons that match. These should all be `>=` comparisons.

Comment (mbhenry) [inline examples/image_ocr.py]: Ok, I see..i hard-coded "20" in the TextGenerator constructor so as submitted it works fine, but if you start changing the start_epoch it doesn't like it.  Like I said, this was a hacky addition. I think using "initial_epoch" I can make it cleaner.

Comment (patyork) [inline examples/image_ocr.py]: Yeah, no worries. It's a great script, and runs fine to completion as-is. It's just a little fragmented when things go wrong and/or training is stopped and restarted.

**Comment by mbhenry** [comment]
OK latest commit uses initial_epoch. Thanks for pointing that out Pat, the curriculum logic is much cleaner now.

**Comment by patyork** [comment]
Looks excellent! Easier to follow and to modify. Stopping/Loading/Starting works perfectly now. And the big thing: I've recreated the results twice. I think it's good to merge.

Just a sidenote - when doing a curriculum like this, I usually use validation loss and total examples seen (in this phase) to determine when to move to the next phase of training, versus the epoch number. In this example, the condition to move to the next phase of training might be `val_los <= -2.0 and examples_seen > 5000`. This prevents hardcoding the epoch numbers, and when done right gets to convergence faster (in time) as when the NN "masters" something, it moves onto the next thing to learn. Doing it in this way handles a lot of variations, like optimizer changes, model changes, `examples_per_epoch` changes, and even when the problem itself changes (e.g. image to voice recognition). Obviously this requires some testing to find the cutoffs, the same as finding the epochs at which to move on. But again, not necessary for this example, just something to keep in mind when making models.

**Comment by fchollet** [comment]
LGTM. Thanks @mbhenry @patyork

**Comment by harikrishnavydana** [comment]
In image_ocr.py
Does the loss that we get in ctc lambada layer mean really the loss or, as written in the code is it some dummy value. @mbhenry @patyork
